tools/jdbc/src/main/java/org/duckdb/DuckDBArrowResult.java

```diff
@@ -0,0 +1,91 @@
+package org.duckdb;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Iterator;
+
+public class DuckDBArrowResult implements Iterable<Object> {
+
+	private Class<?> vector_root_class;
+	private Class<?> schema_class;
+	private Class<?> dictionary_class;
+
+	private Class<?> root_allocator_class;
+	private Class<?> buffer_allocator_class;
+
+	private Class<?> c_schema_class;
+	private Class<?> c_array_class;
+	private Class<?> c_data_class;
+	private Class<?> c_dictionary_class;
+
+	private Object allocator;
+
+	private DuckDBResultSet result_set;
+	private Object schema;
+	private Object vector_schema_root;
+
+	public DuckDBArrowResult(ResultSet result_set_p) throws Exception {
+		if (!(result_set_p instanceof DuckDBResultSet)) {
+			throw new SQLException("This only works with a DuckDB ResultSet");
+		}
+		result_set = (DuckDBResultSet) result_set_p;
+
+		vector_root_class = Class.forName("org.apache.arrow.vector.VectorSchemaRoot");
+		schema_class = Class.forName("org.apache.arrow.vector.types.pojo.Schema");
+		dictionary_class = Class.forName("org.apache.arrow.vector.dictionary.DictionaryProvider");
+
+		root_allocator_class = Class.forName("org.apache.arrow.memory.RootAllocator");
+		buffer_allocator_class = Class.forName("org.apache.arrow.memory.BufferAllocator");
+
+		c_schema_class = Class.forName("org.apache.arrow.c.ArrowSchema");
+		c_array_class = Class.forName("org.apache.arrow.c.ArrowArray");
+		c_data_class = Class.forName("org.apache.arrow.c.Data");
+		c_dictionary_class = Class.forName("org.apache.arrow.c.CDataDictionaryProvider");
+
+		allocator = root_allocator_class.getConstructor().newInstance();
```

Comment:
I believe that the arrow recommendation is to have a single Root Allocator for an application and re-use that root allocator to create "child" allocators. While not necessary, it provides a way to limit the memory allocation for the entire application or a specific use case. I'm unsure how one might pass in / use an application root allocator given the dependencies restrictions in place. 

The allocator should also be closed after use. 

From their [development guidelines](https://arrow.apache.org/docs/java/memory.html#development-guidelines)
> Applications should generally:
> + Use the BufferAllocator interface in APIs instead of RootAllocator.
> + Create one RootAllocator at the start of the program.
> + close() allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.

While this isn't a blocker for me, you may still want to make sure to close this allocator after use to prevent any leaks. I'm not sure if the garbage collection of the allocator correctly avoid memory leaks or if it needs to be closed to release the "direct" memory.

Reply:
The interface requires passing an arrow allocator now

Reply:
👍🏻 Looks good.